source/BookEntry.cpp

```diff
@@ -0,0 +1,161 @@
+/* BookEntry.cpp
+Copyright (c) 2025 by xobes
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "BookEntry.h"
+
+#include "DataNode.h"
+#include "DataWriter.h"
+#include "text/Format.h"
+#include "image/Sprite.h"
+#include "image/SpriteSet.h"
+#include "shader/SpriteShader.h"
+#include "text/WrappedText.h"
+
+using namespace std;
+
+
+
+// Note: A BookEntry exists potentially in advance of having taken effect and being placed into
+// the logbook, e.g. when it exists as merely a potential outcome of a given GameAction.
+// When the GameAction is triggered, the Instantiate method will be called to perform necessary
+// substitutions on the text at that time.
+BookEntry::BookEntry()
+{
+}
+
+
+
+bool BookEntry::Empty() const
+{
+	if(items.empty())
+		return true;
+	for(const ItemType &item : items)
+		if(!std::holds_alternative<std::monostate>(item))
+			return false;
+	return true;
+}
+
+
+
+void BookEntry::Read(const DataNode &node, int startAt)
+{
+	// Note: Like Dialog::ParseTextNode, BookEntry::Read will consume all remaining Tokens of this node
+	// as well as all remaining tokens of the children of this node.
+
+	// First, consume the rest of this first node:
+	if(startAt <= node.Size())
+		AppendItem(ReadItem(node, startAt));
+
+	// Then continue with its child nodes:
+	for(const DataNode &child : node)
+		AppendItem(ReadItem(child));
```

Comment:
These comments aren't particularly useful. They aren't meaningfully shorter than the code they describe and that code is pretty straightforward already.
```suggestion
	if(startAt <= node.Size())
		AppendItem(ReadItem(node, startAt));

	for(const DataNode &child : node)
		AppendItem(ReadItem(child));
```